crates/algorithms/src/haversine/mod.rs

```diff
@@ -0,0 +1,131 @@
+use core::fmt::Debug;
+use nalgebra::Point2;
+use num_traits::Float;
+
+///calculate distance between point with use haversine formula
+/// return distance in km
+#[cfg_attr(
+    feature = "tracing",
+    tracing::instrument("Calculate Haversine Data", skip_all)
+)]
+fn calculate_haversine_distance<T>(p1: Point2<T>, p2: Point2<T>, sphere_radius: T) -> T
+where
+    T: Copy + Float + Debug + 'static,
+{
+    //define two = 2 to support generic
+    let two = T::from(2.0).unwrap();
+    //calc delta
+    let dlat = (p2.x - p1.x).to_radians();
+    let dlon = (p2.y - p1.y).to_radians();
+    //convert coordinate to radians
+    let lat1_radians = p1.x.to_radians();
+    let lat2_radians = p2.x.to_radians();
+    //apply haversine formula
+    let a = (dlat / two).sin().powi(2)
```

Comment:
By restricting T to also have the ClosedSub trait.
We can do something like this:
` 
    let delta = point_b - point_a;

    let lat1_in_rad = point_a.x.to_radians();
    let lat2_in_rad = point_b.x.to_radians();

    let a = squared_rad_sine(delta.x)
        + squared_rad_sine(delta.y) * lat1_in_rad.cos() * lat2_in_rad.cos();

    let c = (T::one() + T::one()) * a.sqrt().atan2((T::one() - a).sqrt());

    sphere_radius * c
`
 
We can add a function before this function and make it inline to ensure it is optimized out
`
#[inline]
fn squared_rad_sine<T>(input: T) -> T
where
    T: Float,
{
    (input.to_radians() / (T::one() + T::one())).sin().powi(2)
}
` 

This way we avoid writing duplicate code and still ensure runtime performance :)